Updated designs for Agency Settings page

ivan-kishko opened the pull request:
## Description of the change

> Updated designs according to figma for agency settings page:
1. Split basic info to readonly part and description editable part
2. Made team, supervisions and jurisdictions editable, fix designs
3. Allow only 1 setting to edit at a time 

## Type of change

Type: Non-breaking refactor

> All pull requests must have at least one of the following labels applied (otherwise the PR will fail):

| Label                       	| Description                                                                                               	|
|-----------------------------	|-----------------------------------------------------------------------------------------------------------	|
| Type: Bug                   	| non-breaking change that fixes an issue                                                                   	|
| Type: Feature               	| non-breaking change that adds functionality                                                               	|
| Type: Breaking Change       	| fix or feature that would cause existing functionality to not work as expected                            	|
| Type: Non-breaking refactor 	| change addresses some tech debt item or prepares for a later change, but does not change functionality    	|
| Type: Configuration Change  	| adjusts configuration to achieve some end related to functionality, development, performance, or security 	|
| Type: Dependency Upgrade      | upgrades a project dependency - these changes are not included in release notes                             |

## Related issues

Closes #244 

## Checklists

### Development

**This box MUST be checked by the submitter prior to merging**:
- [x] **Double- and triple-checked that there is no Personally Identifiable Information (PII) being mistakenly added in this pull request**

These boxes should be checked by the submitter prior to merging:
- [ ] Tests have been written to cover the code changed/added as part of this pull request

### Code review

These boxes should be checked by reviewers prior to merging:

- [ ] This pull request has a descriptive title and information useful to a reviewer
- [ ] This pull request has been moved out of a Draft state, has no "Work In Progress" label, and has assigned reviewers
- [ ] Potential security implications or infrastructural changes have been considered, if relevant

ivan-kishko commented:
Hello to all! Hope You all had nice Holidays!
About PR: 
1. Have not done autocomplete for Jurisdictions feature, and not because of locations data missing but mostly because I still cannot get full picture of what is happening inside Jurisdictions settings.
2. For removing team member I need some clarification on how should confirmation modal appear/look like.

Apart from what I said above and missing api interactions everything is done. Looking forward for your review and comments! Thanks)

hobuobi commented:
In general, this is looking great! At a high-level: 
- Almost all of my comments are actually design comments, not implementation bugs. So I'll CC @agronomic to take a look through these! 
- In reviewing this, I think the "Cancel" functionality is a bit more complex. Going to tackle that in a separate comment. 

### General

- Now I _really_ think we should move the Edit button to the header of the section – it feels strange to scroll all the way to the bottom of the section before you can edit. 
- In all places, “Done” should be “Save” if we’re actually not going to save their changes until they hit the button.
- This isn’t a problem with your implementation, but the padding on the main container makes it such that the content cuts off before the navbar, which feels unnatural. No need to address it in this PR since this applies to other pages too.

### Agency Details (the box)

- “If any of the **above** looks incorrect…”

### Agency Information

- All good, great work!

### Team Management

- The description (“Invite, remove, or assign people as an admin”) is right above the Invite form, which is confusing. Should probably say “Invite users to join your agency on Publisher.”
- Seems like the description “People who have access to Publisher” should actually be more specific, like “Select people to remove or assign Admin status.”
- We should definitely have some kind of hover state for each of the line items.
    - Related: Maybe even make the whole line item clickable / the same thing as hitting the checkmark. The checkmark is too small of a hitbox.
- With large lists, it’s very possible to not see anything in the interface change when you select an item. Maybe we should make some things fixed height or static in order to make sure the user sees the actions at the bottom (“Make Admin”, etc) appear.

### Jurisdictions

- Magnifying glass icon should be static in the search box.
- Not seeing any implementation of the search functionality – we can provide a mock data structure containing some jurisdictions if that would be helpful! Let me know.

hobuobi commented:
### Regarding Canceling/Saving Behavior:

All of this actually depends on an engineering question – **will we actually not save any of their changes until they press save** (in which case the frontend needs to reflect a different state than the backend)?

If that’s true, then:
- “Done” should say “Save” instead.
-  Hitting Cancel should ask for the user’s confirmation (”Are you sure you’d like to exit without saving your changes?”), since it could potentially clear a lot of the information that they’ve entered. 
    - Clicking outside the box should have the same experience.
    - Trying to edit another section should also have a similar experience; ask the user if they want to cancel their edits on the active section, and then if they say yes, exit the current section and open the other section in Edit Mode. (That way we are respecting the user’s choice to edit a new section.)
- I don’t think we need to ask for the user to confirm before removing jurisdictions; this seems redundant.
- The entire Team Management section feels like it requires a different treatment; “inviting” users seems like it should immediately send an email request, and “removing users” feels like it should immediately remove the user.
    - Alternatively: If we decided to keep the same treatment as all other sections (these additions/removals are only saved when the user clicks “Save”), then the Save confirmation modal should summarize the changes (Add X users, remove Y users).

CC: @lilidworkin @agronomic

lilidworkin commented:
> will we actually not save any of their changes until they press save

@hobuobi we can implement this either way you'd like! From looking at the designs, it seems to me that the most natural behavior is:
* Agency information section: save when the user clicks "Save"
* Team management section: saves happen immediately when you click "Invite" or "Remove from Agency". Clicking "Done" just exits the modal and doesn't do anything (in which case we don't need the "Cancel" button)
* Supervision populations: save happens when the user clicks "Save"
* Jurisdictions section: save happens when the user clicks "Save"

But we can adjust any of this behavior depending on product/design preferences!

ivan-kishko commented:
Hello to all. @hobuobi I made some static changes you addressed, like changed Done to save, fixed typo in basics, changed text in teams, made container for members fixed with scroll in edit mode, made hover state for rows, made search icon static even with text and mocked autocomplete flow for now with mocks I made myself. 

Regarding other changes: I probably need design for moving edit button to the top - like where exactly, If you want Cancel be complex, then need design for this confirmations too, (not sure why you want treat click anywhere outside edit box like Cancel action, it would be like even if you click unintentionally you will get a prompt modal with confirmation).   

> I don’t think we need to ask for the user to confirm before removing jurisdictions; this seems redundant.

not sure what this is about...

And about Team management flow - I guess when we have API you can make final decision on how it should be.

lilidworkin commented:
@ivan-kishko this is looking amazing, great work!! I believe we will be getting answers to the design questions soon, so we can wait on those for now. A few of my own thoughts:

## Design questions

@ivan-kishko you don't need to address these, putting them here for @agronomic .

The empty state (when you have no description, no team members, and no supervision populations) looks strange to me: 
<img width="501" alt="Screenshot 2023-01-05 at 11 34 30 AM" src="https://user-images.githubusercontent.com/99899591/210833548-81ea5210-409e-478d-a350-9987fca6f0f6.png">
The "no supervision populations" looks especially strange, because the lack of any icon makes it seem like you're saying you *can* disaggregate by all those items.

When you have a long list of team members, it looks unwieldy, should we add a fixed height + scroll?
<img width="527" alt="Screenshot 2023-01-05 at 11 41 59 AM" src="https://user-images.githubusercontent.com/99899591/210833744-ff10c499-24f3-4ff1-910b-9219866f53a4.png">

## Engineering feedback
I noticed one bug (maybe?) where it seems you can select supervision populations in view mode:
https://user-images.githubusercontent.com/99899591/210834057-f46b7381-0f18-405b-b094-dc9987016fae.mov

I love the red highlight animation when you try to edit more than one section at once!!

I know we don't have backend functionality yet, but I wonder if we can still update the frontend state when you add/remove team members or jurisdictions, so we can see how it feels? Obviously those changes wouldn't persist if you load the page. Same with clicking on jurisdictions after you search for them -- would be great to see the list updated after that.

We still need to implement the "exclude jurisdictions" flow. @hobuobi can you clarify your comment about not needing confirmation for this flow? I don't see a confirmation in the designs. Should we still implement the flow I see in Figma [here](https://www.figma.com/file/dvX3yg7FMGJSWFaqHWrWgQ/Justice-Counts-Publisher?node-id=6255%3A15227&t=rofwy0w8tCTyYsyO-0)?

terryttsai commented on publisher/src/components/Settings/AgencySettingsBasicInfo.tsx:
```diff
@@ -0,0 +1,57 @@
+        {currentAgency?.state}
+      </BasicInfoRow>
+      <BasicInfoRow>
+        <span>Systems</span>
```
Maybe this should say "System" instead of "Systems" if there's only one system?

ivan-kishko replied:
done [4fe42f5](https://github.com/Recidiviz/justice-counts/pull/264/commits/4fe42f507f74deef5bd20a46dfe80406664fb4fe)

terryttsai commented on publisher/src/components/Settings/AgencySettings.styles.tsx:
```diff
@@ -19,84 +19,125 @@ import {
 
 // 662px is settings menu width times 2
-// 664px is the width of agency settings container
+// 732px is the width of agency settings container
```
We could define 732 as a constant and reuse it?

ivan-kishko replied:
[4fe42f5](https://github.com/Recidiviz/justice-counts/pull/264/commits/4fe42f507f74deef5bd20a46dfe80406664fb4fe)

terryttsai commented on publisher/src/components/Settings/AgencySettings.tsx:
```diff
@@ -15,101 +15,85 @@
+  Description = "DESCRIPTION",
+  Team = "TEAM",
+  Supervisions = "SUPERVISIONS",
+  Jurisdictions = "Jurisdictions",
```
Should this be `Jurisdictions = "JURISDICTIONS"` for consistency?

ivan-kishko replied:
[4fe42f5](https://github.com/Recidiviz/justice-counts/pull/264/commits/4fe42f507f74deef5bd20a46dfe80406664fb4fe), my bad

terryttsai commented on publisher/src/components/Settings/SettingsMenu.tsx:
```diff
@@ -109,27 +92,27 @@ export const SettingsMenu: React.FC = observer(() => {
-                ))}
-              </SubMenuListContainer>
-            )}
+          {/* {removeAgencyFromPath(location.pathname) === */}
```
Should this comment be removed?

ivan-kishko replied:
[4fe42f5](https://github.com/Recidiviz/justice-counts/pull/264/commits/4fe42f507f74deef5bd20a46dfe80406664fb4fe)

terryttsai commented on publisher/src/components/Settings/AgencySettings.tsx:
```diff
@@ -15,101 +15,85 @@
+  isSettingInEditMode: boolean;
+  openSetting: () => void;
+  closeSetting: () => void;
+  showAnimation: boolean;
```
nit: I like to phrase booleans like `isAnimationShowing` because `showAnimation` to me sounds like a function you can invoke

ivan-kishko replied:
[4fe42f5](https://github.com/Recidiviz/justice-counts/pull/264/commits/4fe42f507f74deef5bd20a46dfe80406664fb4fe), renamed other similar variables as well

terryttsai commented on publisher/src/components/Settings/AgencySettings.tsx:
```diff
@@ -15,101 +15,85 @@
+    setActiveSetting(undefined);
+    setShowActiveSettingAnimation(false);
+  };
+  const generateSettingProps = (settingName: ActiveSetting): SettingProps => ({
```
Love this generator!

mxosman replied:
Agreed! Borrowing this idea for the future.

terryttsai commented on publisher/src/components/Settings/AgencySettingsTeamManagement.tsx:
```diff
@@ -120,54 +180,58 @@ export const AgencySettingsTeamManagement: React.FC<{
+              value={emailValue}
+              onChange={(e) => setEmailValue(e.target.value)}
+            />
+            <InviteMemberButton
```
Could disabled the invite button until all fields are filled out and valid?

ivan-kishko replied:
Well, there is no way for a frontend to ensure by regex that email is correct valid and belongs to a user. If you want to do that you have to decide what pattern you consider to be valid for a front end validation and I'll do the rest.

ivan-kishko replied:
for now just made a check so this fields are not empty

terryttsai replied:
That sounds good to me! Thanks for doing that

lilidworkin replied:
Agreed, this is fine!! Good suggestion @terryttsai  and thanks @ivan-kishko !

terryttsai reviewed (approved):
Looking good, left a small suggestion!

hobuobi commented:
> @hobuobi we can implement this either way you'd like! From looking at the designs, it seems to me that the most natural behavior is:
> 
> * Agency information section: save when the user clicks "Save"
> * Team management section: saves happen immediately when you click "Invite" or "Remove from Agency". Clicking "Done" just exits the modal and doesn't do anything (in which case we don't need the "Cancel" button)
> * Supervision populations: save happens when the user clicks "Save"
> * Jurisdictions section: save happens when the user clicks "Save"

This is the most natural behavior to me as well! In that case, I think we can go with the suggestions in my previous comment: 
- Cancel should ask the user for confirmation before clearing their progress ("Are you sure you would like to cancel your changes?"): 
    - If they select No, return to the editMode box.
    - If they select Yes, don't save their changes and go back to the default Agency Settings page (editMode off for all) OR to editMode for whichever section the user clicked edit for (see below). 
- The following should all trigger the Cancel function: 
    - Clicking Cancel.
    - Clicking outside the box. 
    - Clicking editMode for a different section.

mxosman commented on agency-dashboard/src/Header/AboutModal.tsx:
```diff
@@ -48,12 +51,14 @@ export const AboutModal: React.FC<{
+          <AboutModalButtonsContainer
+            onClick={() => {
+              if (agencyUrl) {
+                window.open(formatExternalLink(agencyUrl), "_blank");
```
nit: I would pass in the third argument of 'noreferrer' just to be safe. Could be done in the next PR(s)!

`window.open(formatExternalLink(agencyUrl), "_blank", "noreferrer");`

ivan-kishko replied:
ignoring, see Lili's below comment

mxosman commented on publisher/src/components/Settings/AgencySettingsDescription.tsx:
```diff
@@ -0,0 +1,128 @@
+            maxLength={750}
+          />
+          <AgencyInfoTextAreaWordCounter isRed={infoText.length >= 750}>
+            {infoText.length}/750 characters
```
tiny nit: not a big deal - but could make 750 as a constant to make it easier to change the max character limit in the future?

ivan-kishko replied:
[083f738](https://github.com/Recidiviz/justice-counts/pull/264/commits/083f7384c9aeaa83496fb595fba8b21ff563e909)

mxosman commented:
This feels great, Ivan! One small thing in Team Management - I am able to make someone an admin, but I cannot remove them from being an admin, should there be functionality to remove admin status (apologies if this has already been discussed)? 

Otherwise - just had a couple of small nits!

ivan-kishko commented:
> This feels great, Ivan! One small thing in Team Management - I am able to make someone an admin, but I cannot remove them from being an admin, should there be functionality to remove admin status (apologies if this has already been discussed)?
> 
> Otherwise - just had a couple of small nits!

Hello @mxosman!) About removing admin rights - as far as I can see on design there is no such flow, either you cannot do that and can only remove a user with admin rights or this is not currently on design yet.
About comments, AgencySettingsURL and AboutModal are in my PR only because of merge with other branch I guess, I have not created this code and If it is ok I can just rely on your comments to make those fixes.

ivan-kishko commented on publisher/src/components/Settings/AgencySettingsURL.tsx:
```diff
@@ -0,0 +1,125 @@
+          <AgencyInfoBlockDescription>
+            <AgencyInfoLink
+              href={formatExternalLink(settings.HOMEPAGE_URL)}
+              target="_blank"
```
ignoring, see Lili's below comment

ivan-kishko commented:
@lilidworkin @hobuobi Hello!) I made a commit for this whole "cancel setting changes/open another setting" flow, check this out please and if all good I am gonna merge. (P.S. Modal looks awful, probably in near future you make a design for it as well and I fix it)

mxosman commented:
> Hello @mxosman!) About removing admin rights - as far as I can see on design there is no such flow, either you cannot do that and can only remove a user with admin rights or this is not currently on design yet. 

No problem at all! Sounds great! None of my comments are blocking at all - this is fantastic! Thank you, Ivan!
